lib/spack/spack/stage.py

```diff
@@ -45,18 +75,16 @@ def _first_accessible_path(paths):
                 if can_access(path):
                     return path
             else:
-                # The path doesn't exist so create it and adjust permissions
-                # and group as needed (e.g., shared ``$tempdir``).
-                prefix = os.path.sep
-                parts = path.strip(os.path.sep).split(os.path.sep)
-                for part in parts:
-                    prefix = os.path.join(prefix, part)
-                    if not os.path.exists(prefix):
-                        break
-                parent = os.path.dirname(prefix)
-                gid = os.stat(parent).st_gid
-                mkdirp(path, group=gid, default_perms='parents')
-
+                # Remove adjacent occurance of `$user` (e.g., `$tempdir/$user`
+                # where `$tempdir` includes `$user`).
+                user = getpass.getuser()
+                user_node = '{0}{1}'.format(os.path.sep, user)
+                double_user = '{0}{0}'.format(user_node)
+                if double_user in path:
+                    path = path.replace(double_user, user_node, 1)
+
+                # Now create the stage root with the proper group/perms.
+                _create_stage_root(path)
                 if can_access(path):
```

Comment:
I am hesitant to remove the check, which was essentially there pre-12072, in part because it may still come in handy with parallel/distributed builds involving a shared or service account.

Reply:
> in part because it may still come in handy with parallel/distributed builds

If we're in this part of the if/else block, it means the path didn't exist; if the creation function succeeds without an exception, it means we created the file (and have permissions for it).

I'm not sure what this check would add for parallel builds. While I think it's worth checking the pre-12072 code (and associated commits) I think the biggest factor in whether this check should persist is whether it makes sense in the context of the current code.

Reply:
What happens if another process creates the directory between the original check and the call to `_create_stage_root`?  Both `_create_stage_root` and `mkdirp` skip directories that you can check for existence.

Reply:
I think I see how that might happen with `_create_stage_root` as it is written: it checks `if os.path.exists` before creating each subdirectory; an interleaving of processes could cause an issue. 

I think it would be clearer if `_create_stage_root` tried `can_access` itself and throws an exception if it cant: this would IMO make it more idiomatic with other python use cases where you want to create a directory in a multi-user environment (e.g. people complaining about `if not os.path.exists(x): os.mkdir(x)` in https://stackoverflow.com/questions/18973418/os-mkdirpath-returns-oserror-when-directory-does-not-exist/22718321).

Overall admittedly that has the same result but I think it is easier to reason about.

Reply:
I understand your point.

Even if I check `can_access` down to the `$user` directory, `mkdirp` would need to be changed in a similar manner  to ensure `user_subdirs` is handled consistently.

Ideally I'd be using `mkdirp` for the whole process, but it does not handle the arbitrarily located `$user` node permissions case Todd (@tgamblin) emphasized to me in our discussions.  (I'm hoping such a change is not part of this PR since `mkdirp` has wide-ranging use beyond the scope of creating the stage path root.)

Reply:
I wonder why this is not "Outdated" since I replaced the code up to line 84.

Reply:
The code it relates to (`can_access`) is still here, and in the same context. I think `can_access` should move to the end of `_create_stage_root` and throw an exception if it can't

Reply:
So are you saying I should change `mkdirp` to do the same in this PR?

Reply:
No if possible `mkdirp` should remain the same. I think `mkdirp` will generally raise an exception in the parallel case (unlike `_create_stage_root` as it is currently written) because it uses `os.makedirs`, which will throw an `OSError (17)` if the user tries to create a directory that already exists.